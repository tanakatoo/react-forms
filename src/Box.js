import React from "react"

const Box = ({ width, height, color }) => {

    return (
        <div data-testid="box" style={{ width: parseInt(width), height: parseInt(height), backgroundColor: color }}>

        </div>
    )
}

export default Box;