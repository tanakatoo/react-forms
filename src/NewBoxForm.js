import React, { useState } from "react"

const NewBoxForm = ({ addItem }) => {
    const INITIAL_STATE = { width: '', height: '', color: '' }
    const [formData, setFormData] = useState(INITIAL_STATE)

    const handleChange = (e) => {
        const { name, value } = e.target

        setFormData(formData => ({
            ...formData,
            [name]: value
        }))

    }

    const handleSubmit = (e) => {
        e.preventDefault()
        //call something that adds that item into state, this is defined in the BoxList becuase we are adding to BoxList
        //form state is local but we take values and pass values to parent
        addItem(formData.width, formData.height, formData.color)
        console.log(formData.width, formData.height, formData.color)
        setFormData(INITIAL_STATE)

    }

    return (
        <div>
            <form onSubmit={handleSubmit}>
                <label htmlFor="width" >Width</label>
                <input name="width" id="width" type="number" value={formData.width} onChange={handleChange} />
                <label htmlFor="height" >Height</label>
                <input name="height" id="height" type="number" value={formData.height} onChange={handleChange} />
                <label htmlFor="color" >Color</label>
                <input name="color" id="color" type="text" value={formData.color} onChange={handleChange} />
                <button>Add</button>
            </form>
        </div>
    )
}

export default NewBoxForm