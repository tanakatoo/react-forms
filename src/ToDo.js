import React from "react";

const ToDo = ({ task, id, removeTask }) => {

    return (
        <li id={id}>{task} <button onClick={removeTask}>X</button></li>
    )


}

export default ToDo