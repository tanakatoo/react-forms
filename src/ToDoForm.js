import React, { useState } from "react";
import { v4 as uuid } from "uuid"

const ToDoForm = ({ addData }) => {
    const INITIAL_STATE = {
        task: ""
    }
    const [formData, setFormData] = useState(INITIAL_STATE)
    const handleChange = (e) => {
        const { name, value } = e.target
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))

    }

    const handleSubmit = (e) => {
        e.preventDefault()
        addData(formData.task, uuid())
        setFormData(INITIAL_STATE)
    }
    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="task">Task</label>
            <input type='text' name="task" id="task" onChange={handleChange} value={formData.task} />
            <button>Add</button>
        </form>
    )
}

export default ToDoForm