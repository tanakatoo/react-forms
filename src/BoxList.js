import React, { useState } from "react"
import Box from "./Box"
import NewBoxForm from "./NewBoxForm"
import { v4 as uuid } from 'uuid';

const BoxList = (props) => {
    const INITIAL_STATE = []
    const [boxesData, setBoxesData] = useState(INITIAL_STATE)


    const addItem = (width, height, color) => {
        setBoxesData(boxesData => [...boxesData, { width, height, color }])
        console.log(boxesData)
    }


    return (
        <div>
            <NewBoxForm addItem={addItem} />
            <h1>Box List</h1>
            {boxesData.map(item => <Box width={item.width}
                height={item.height}
                color={item.color}
                key={uuid()}
            />)}
        </div>
    )
}

export default BoxList