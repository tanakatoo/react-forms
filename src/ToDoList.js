import React, { useState } from "react"
import ToDoForm from "./ToDoForm"
import ToDo from './ToDo'
import { v4 as uuid } from "uuid"


const ToDoList = () => {
    const INITIAL_STATE = []
    const [toDo, setToDo] = useState(INITIAL_STATE)

    const addData = (task, id) => {
        setToDo(toDo => [...toDo, { task: task, id: id }])
    }

    const removeTask = (e) => {

        setToDo(toDo => toDo.filter(t => t.id != e.target.parentElement.id))
    }
    return (
        <div>
            <ToDoForm addData={addData} />
            <ul>
                {toDo.map(t => <ToDo task={t.task} key={t.id} id={t.id} removeTask={removeTask} />)}
            </ul>
        </div>

    )
}

export default ToDoList