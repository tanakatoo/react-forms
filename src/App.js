import logo from './logo.svg';
import './App.css';
import BoxList from './BoxList'
import ToDoList from './ToDoList'

function App() {
  return (
    <div className="App">
      <BoxList />
      <ToDoList />
    </div>
  );
}

export default App;
