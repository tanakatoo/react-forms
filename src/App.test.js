import { fireEvent, render, screen } from '@testing-library/react';
import App from './App';
import Box from './Box'
import BoxList from './BoxList';
import NewBoxForm from './NewBoxForm';

import ToDoList from './ToDoList';
import ToDo from './ToDo'
import ToDoForm from './ToDoForm';

test('renders Box', () => {
  render(<Box width="30" height="40" color="red" />);
});

test('renders BoxList', () => {
  render(<BoxList />);
});

test('renders NewBoxForm', () => {
  render(<NewBoxForm />);
});

test('if Box matches snapshot', () => {
  const { asFragment } = render(<Box width="30" height="40" color="red" />)
  expect(asFragment()).toMatchSnapshot();
})

test('if BoxList matches snapshot', () => {
  const { asFragment } = render(<BoxList boxesData={[{
    width: 30,
    height: 40,
    color: "red"
  }]} />)
  expect(asFragment()).toMatchSnapshot();
})

test('if Form matches snapshot', () => {
  const { asFragment } = render(<NewBoxForm />)
  expect(asFragment()).toMatchSnapshot();
})

test('if box can be added on the screen', () => {
  const { getByLabelText, queryByText, queryAllByTestId } = render(<BoxList />)

  const inputWidth = getByLabelText('Width');
  const inputHeight = getByLabelText('Height');
  const inputColor = getByLabelText('Color');
  const btn = queryByText("Add")

  expect(queryAllByTestId('box').length).toEqual(0)

  fireEvent.change(inputWidth, { target: { value: 100 } })
  fireEvent.change(inputHeight, { target: { value: 300 } })
  fireEvent.change(inputColor, { target: { value: "yellow" } })
  fireEvent.click(btn)

  const { getAllByTestId } = render(<BoxList />)
  expect(getAllByTestId('box')[0].style.backgroundColor).toEqual("yellow")
  expect(getAllByTestId('box')[0].style.width).toEqual("100px")
  expect(getAllByTestId('box')[0].style.height).toEqual("300px")
})

test('renders ToDoList', () => {
  render(<ToDoList />)
})

test('renders ToDo', () => {
  render(<ToDo />)
})

test('renders ToDoForm', () => {
  render(<ToDoForm />)
})

test('ToDoList same as fragment', () => {
  const { asFragment } = render(<ToDoList />)

  expect(asFragment()).toMatchSnapshot()
})

test('ToDo same as fragment', () => {
  const { asFragment } = render(<ToDo />)

  expect(asFragment()).toMatchSnapshot()
})

test('ToDoForm same as fragment', () => {
  const { asFragment } = render(<ToDoForm />)

  expect(asFragment()).toMatchSnapshot()
})

test('adding a task', () => {
  const { getByLabelText, queryByText } = render(<ToDoList />)

  const input = getByLabelText('Task')
  const btn = queryByText('Add')

  fireEvent.change(input, { target: { value: "get eggs" } })
  fireEvent.click(btn)

  expect(queryByText('get eggs')).toBeInTheDocument()
  expect(queryByText("X")).toBeInTheDocument()

})

test('deleting a task', () => {
  const { getByLabelText, queryByText } = render(<ToDoList />)

  const input = getByLabelText('Task')
  const btn = queryByText('Add')

  fireEvent.change(input, { target: { value: "get eggs" } })
  fireEvent.click(btn)

  const deleteBtn = queryByText("X")

  expect(queryByText('get eggs')).toBeInTheDocument()
  expect(deleteBtn).toBeInTheDocument()

  fireEvent.click(deleteBtn)

  expect(queryByText('get eggs')).not.toBeInTheDocument()


})