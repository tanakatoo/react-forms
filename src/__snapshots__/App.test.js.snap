// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo same as fragment 1`] = `
<DocumentFragment>
  <li>
     
    <button>
      X
    </button>
  </li>
</DocumentFragment>
`;

exports[`ToDoForm same as fragment 1`] = `
<DocumentFragment>
  <form>
    <label
      for="task"
    >
      Task
    </label>
    <input
      id="task"
      name="task"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
</DocumentFragment>
`;

exports[`ToDoList same as fragment 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        Task
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;

exports[`if Box matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="box"
    style="width: 30px; height: 40px; background-color: red;"
  />
</DocumentFragment>
`;

exports[`if BoxList matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          type="number"
          value=""
        />
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          type="number"
          value=""
        />
        <label
          for="color"
        >
          Color
        </label>
        <input
          id="color"
          name="color"
          type="text"
          value=""
        />
        <button>
          Add
        </button>
      </form>
    </div>
    <h1>
      Box List
    </h1>
  </div>
</DocumentFragment>
`;

exports[`if Form matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
